dynamic-check/src/objects/frame.rs

```diff
@@ -0,0 +1,1198 @@
+    /// Creates a new membership type assertion between a given and the set of index types.
+    ///
+    fn assert_index(&mut self, given: &Type, span: &Span) {
+        let index_types = Type::index_types();
+
+        self.assert_membership(given.clone(), index_types, span)
+    }
+
+    ///
+    /// Collects a vector of `TypeAssertion` predicates from a vector of statements.
+    ///
+    fn parse_statements(&mut self) -> Result<(), FrameError> {
+        for statement in self.statements.clone() {
+            self.parse_statement(&statement)?;
+        }
+
+        Ok(())
+    }
+
+    ///
+    /// Collects a vector of `TypeAssertion` predicates from a statement.
+    ///
+    fn parse_statement(&mut self, statement: &Statement) -> Result<(), FrameError> {
+        match statement {
+            Statement::Return(expression, span) => self.parse_return(expression, span),
+            Statement::Definition(declare, variables, expression, span) => {
+                self.parse_definition(declare, variables, expression, span)
+            }
+            Statement::Assign(assignee, expression, span) => self.parse_assign(assignee, expression, span),
+            Statement::Conditional(conditional, span) => self.parse_statement_conditional(conditional, span),
+            Statement::Iteration(identifier, from_to, statements, span) => {
+                self.parse_iteration(identifier, from_to, statements, span)
+            }
+            Statement::Expression(expression, span) => self.parse_statement_expression(expression, span),
+            Statement::Console(_console_call) => Ok(()), // Console function calls do not generate type assertions.
+        }
+    }
+
+    ///
+    /// Collects `TypeAssertion` predicates from a return statement.
+    ///
+    fn parse_return(&mut self, expression: &Expression, span: &Span) -> Result<(), FrameError> {
+        // Get the function output type.
+        let output_type = &self.function_type.output.type_;
+
+        // Create the left hand side of a type assertion.
+        let left = output_type.clone();
+
+        // Create the right hand side from the statement return expression.
+        let right = self.parse_expression(expression)?;
+
+        // Create a new type assertion for the statement return.
+        self.assert_equal(left, right, span);
+
+        Ok(())
+    }
+
+    ///
+    /// Collects `Type Assertion` predicates from a definition statement.
+    ///
+    fn parse_definition(
+        &mut self,
+        _declare: &Declare,
+        variables: &Variables,
+        expression: &Expression,
+        span: &Span,
+    ) -> Result<(), FrameError> {
+        // Parse the definition expression.
+        let actual_type = self.parse_expression(expression)?;
+
+        // Check if an explicit type is given.
+        if let Some(type_) = variables.type_.clone() {
+            // Convert the expected type into a dynamic check type.
+            let expected_type = match self.self_type.clone() {
```

Comment:
since only the `circuit_type`'s identifier is later passed by value, I'd defer cloning to that point (instead of cloning the whole `circuit_type` here)